
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Predict Effects of Alzheimer’s Variants Using ATAC Models &#8212; gReLU</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=3ba690ee" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/graphviz.css?v=eafc0fe6" />
  
  <!-- So that users can add custom icons -->
  <script src="../_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="../_static/documentation_options.js?v=4e48d009"></script>
    <script src="../_static/doctools.js?v=888ff710"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <link rel="canonical" href="https://genentech.github.io/gReLU/tutorials/5_variant.html" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Querying the public gReLU model zoo on Weights and Biases (wandb)" href="6_model_zoo.html" />
    <link rel="prev" title="Model-guided design of regulatory elements" href="4_design.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="unknown" />
    <meta name="docbuild:last-update" content="Aug 26, 2025"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  
    
  

<a class="navbar-brand logo" href="../index.html">
  
  
  
  
  
  
    <p class="title logo__title">gReLU</p>
  
</a></div>
    
      <div class="navbar-item">
<div class="version-switcher__container dropdown pst-js-only">
  <button id="pst-version-switcher-button-2"
    type="button"
    class="version-switcher__button btn btn-sm dropdown-toggle"
    data-bs-toggle="dropdown"
    aria-haspopup="listbox"
    aria-controls="pst-version-switcher-list-2"
    aria-label="Version switcher list"
  >
    Choose version  <!-- this text may get changed later by javascript -->
    <span class="caret"></span>
  </button>
  <div id="pst-version-switcher-list-2"
    class="version-switcher__menu dropdown-menu list-group-flush py-0"
    role="listbox" aria-labelledby="pst-version-switcher-button-2">
    <!-- dropdown will be populated by javascript on page load -->
  </div>
</div></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../readme.html">
    Overview
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="1_inference.html">
    Tutorial: Inference
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="2_finetune.html">
    Tutorial: Fine-tuning
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="3_train.html">
    Tutorial: Training
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="4_design.html">
    Tutorial: Design
  </a>
</li>

            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button"
                data-bs-toggle="dropdown" aria-expanded="false"
                aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
<li class=" current active">
  <a class="nav-link dropdown-item nav-internal" href="#">
    Tutorial: Variant effect prediction
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="6_model_zoo.html">
    Tutorial: Model zoo
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../contributing.html">
    Contributing
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../license.html">
    License
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../authors.html">
    Contributors
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../changelog.html">
    Changelog
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../autoapi/index.html">
    API Reference
  </a>
</li>

                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/Genentech/gReLU" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fab fa-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../readme.html">
    Overview
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="1_inference.html">
    Tutorial: Inference
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="2_finetune.html">
    Tutorial: Fine-tuning
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="3_train.html">
    Tutorial: Training
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="4_design.html">
    Tutorial: Design
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="#">
    Tutorial: Variant effect prediction
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="6_model_zoo.html">
    Tutorial: Model zoo
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../contributing.html">
    Contributing
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../license.html">
    License
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../authors.html">
    Contributors
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../changelog.html">
    Changelog
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../autoapi/index.html">
    API Reference
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/Genentech/gReLU" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fab fa-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"></div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">Predict Effects of Alzheimer’s Variants Using ATAC Models</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="predict-effects-of-alzheimer-s-variants-using-atac-models">
<h1>Predict Effects of Alzheimer’s Variants Using ATAC Models<a class="headerlink" href="#predict-effects-of-alzheimer-s-variants-using-atac-models" title="Link to this heading">#</a></h1>
<p>In this tutorial, we demonstrate how to predict the impact of sequence variation using a trained gReLU model.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">anndata</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ad</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
</pre></div>
</div>
</div>
</div>
<section id="load-the-catlas-model">
<h2>Load the CATlas model<a class="headerlink" href="#load-the-catlas-model" title="Link to this heading">#</a></h2>
<p>This is a binary classification model trained on snATAC-seq data from Catlas (<a class="reference external" href="http://catlas.org/humanenhancer/">http://catlas.org/humanenhancer/</a>). This model predicts the probability that an input sequence will be accessible in various cell types.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">grelu.resources</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">grelu</span><span class="o">.</span><span class="n">resources</span><span class="o">.</span><span class="n">load_model</span><span class="p">(</span><span class="n">project</span><span class="o">=</span><span class="s1">&#39;human-atac-catlas&#39;</span><span class="p">,</span> <span class="n">model_name</span><span class="o">=</span><span class="s1">&#39;model&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
</section>
<section id="view-the-model-s-metadata">
<h2>View the model’s metadata<a class="headerlink" href="#view-the-model-s-metadata" title="Link to this heading">#</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">model.data_params</span></code> is a dictionary containing metadata about the data used to train the model. Let’s look at what information is stored:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">model</span><span class="o">.</span><span class="n">data_params</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>dict_keys([&#39;tasks&#39;, &#39;train&#39;, &#39;val&#39;, &#39;test&#39;])
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">model</span><span class="o">.</span><span class="n">data_params</span><span class="p">[</span><span class="s1">&#39;train&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="k">if</span> <span class="n">k</span> <span class="o">!=</span><span class="s2">&quot;intervals&quot;</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="n">model</span><span class="o">.</span><span class="n">data_params</span><span class="p">[</span><span class="s1">&#39;train&#39;</span><span class="p">][</span><span class="n">k</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>bin_size 1
end both
genome hg38
label_aggfunc None
label_len 200
label_transform_func None
max_label_clip None
max_pair_shift 0
max_seq_shift 2
min_label_clip None
n_alleles 1
n_augmented 1
n_seqs 977014
n_tasks 204
padded_label_len 200
padded_seq_len 204
predict False
rc True
seq_len 200
</pre></div>
</div>
</div>
</div>
<p>Note the parameter <code class="docutils literal notranslate"><span class="pre">seq_len</span></code>. This tells us that the model was trained on 200 bp long sequences.</p>
<p><code class="docutils literal notranslate"><span class="pre">model.data_params['tasks']</span></code> is a large dictionary containing metadata about the output tracks that the model predicts. We can collect these into a dataframe called tasks:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tasks</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">model</span><span class="o">.</span><span class="n">data_params</span><span class="p">[</span><span class="s1">&#39;tasks&#39;</span><span class="p">])</span>
<span class="n">tasks</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>cell type</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Follicular</td>
      <td>Follicular</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Fibro General</td>
      <td>Fibro General</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Acinar</td>
      <td>Acinar</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</section>
<section id="load-alzheimer-s-variants-from-gwas-catalog-jansen-et-al-2019-meta-analysis">
<h2>Load Alzheimer’s Variants from GWAS Catalog (Jansen et al. 2019 meta-analysis)<a class="headerlink" href="#load-alzheimer-s-variants-from-gwas-catalog-jansen-et-al-2019-meta-analysis" title="Link to this heading">#</a></h2>
<p>Download a small subset of variants from the AD sumstats file from this meta-analysis study. This contains 1,000 variants mapped to the hg19 genome.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">grelu.resources</span>

<span class="n">variant_dir</span> <span class="o">=</span> <span class="n">grelu</span><span class="o">.</span><span class="n">resources</span><span class="o">.</span><span class="n">get_artifact</span><span class="p">(</span>
    <span class="n">project</span><span class="o">=</span><span class="s1">&#39;alzheimers-variant-tutorial&#39;</span><span class="p">,</span>
    <span class="n">name</span><span class="o">=</span><span class="s1">&#39;dataset&#39;</span>
<span class="p">)</span><span class="o">.</span><span class="n">download</span><span class="p">()</span>

<span class="n">variant_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">variant_dir</span><span class="p">,</span> <span class="s2">&quot;variants.txt&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">variants</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="n">variant_file</span><span class="p">)</span>
<span class="n">variants</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>snpid</th>
      <th>chrom</th>
      <th>pos</th>
      <th>alt</th>
      <th>ref</th>
      <th>rsid</th>
      <th>zscore</th>
      <th>pval</th>
      <th>nsum</th>
      <th>neff</th>
      <th>direction</th>
      <th>eaf</th>
      <th>beta</th>
      <th>se</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>6:32630634_G_A</td>
      <td>chr6</td>
      <td>32630634</td>
      <td>G</td>
      <td>A</td>
      <td>6:32630634</td>
      <td>3.974476</td>
      <td>0.000071</td>
      <td>71639</td>
      <td>71639.0</td>
      <td>?+?+</td>
      <td>0.2237</td>
      <td>0.025194</td>
      <td>0.006339</td>
    </tr>
    <tr>
      <th>1</th>
      <td>6:32630797_A_G</td>
      <td>chr6</td>
      <td>32630797</td>
      <td>A</td>
      <td>G</td>
      <td>6:32630797</td>
      <td>4.040244</td>
      <td>0.000053</td>
      <td>71639</td>
      <td>71639.0</td>
      <td>?+?+</td>
      <td>0.2435</td>
      <td>0.024866</td>
      <td>0.006155</td>
    </tr>
    <tr>
      <th>2</th>
      <td>6:32630824_T_C</td>
      <td>chr6</td>
      <td>32630824</td>
      <td>T</td>
      <td>C</td>
      <td>6:32630824</td>
      <td>3.921736</td>
      <td>0.000088</td>
      <td>71639</td>
      <td>71639.0</td>
      <td>?+?+</td>
      <td>0.1859</td>
      <td>0.026630</td>
      <td>0.006790</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</section>
<section id="filter-variants">
<h2>Filter variants<a class="headerlink" href="#filter-variants" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">grelu.data.preprocess</span><span class="w"> </span><span class="kn">import</span> <span class="n">filter_blacklist</span><span class="p">,</span> <span class="n">filter_chromosomes</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">grelu.variant</span><span class="w"> </span><span class="kn">import</span> <span class="n">filter_variants</span>
</pre></div>
</div>
</div>
</div>
<p>Remove indels, since we don’t support them for now. We also remove variants where one of the alleles contains Ns.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">variants</span> <span class="o">=</span> <span class="n">filter_variants</span><span class="p">(</span><span class="n">variants</span><span class="p">,</span> <span class="n">max_del_len</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">max_insert_len</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">standard_bases</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Initial number of variants: 1000
Final number of variants: 989
</pre></div>
</div>
</div>
</div>
<p>Remove non-standard chromosomes</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">variants</span> <span class="o">=</span> <span class="n">filter_chromosomes</span><span class="p">(</span><span class="n">variants</span><span class="p">,</span> <span class="n">include</span><span class="o">=</span><span class="s1">&#39;autosomesXY&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Keeping 989 intervals
</pre></div>
</div>
</div>
</div>
<p>Remove SNPs from unmappable regions</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">variants</span> <span class="o">=</span> <span class="n">filter_blacklist</span><span class="p">(</span><span class="n">variants</span><span class="p">,</span> <span class="n">genome</span><span class="o">=</span><span class="s2">&quot;hg19&quot;</span><span class="p">,</span> <span class="n">window</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Keeping 988 intervals
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">variants</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>snpid</th>
      <th>chrom</th>
      <th>pos</th>
      <th>alt</th>
      <th>ref</th>
      <th>rsid</th>
      <th>zscore</th>
      <th>pval</th>
      <th>nsum</th>
      <th>neff</th>
      <th>direction</th>
      <th>eaf</th>
      <th>beta</th>
      <th>se</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>6:32630634_G_A</td>
      <td>chr6</td>
      <td>32630634</td>
      <td>G</td>
      <td>A</td>
      <td>6:32630634</td>
      <td>3.974476</td>
      <td>0.000071</td>
      <td>71639</td>
      <td>71639.0</td>
      <td>?+?+</td>
      <td>0.2237</td>
      <td>0.025194</td>
      <td>0.006339</td>
    </tr>
    <tr>
      <th>1</th>
      <td>6:32630797_A_G</td>
      <td>chr6</td>
      <td>32630797</td>
      <td>A</td>
      <td>G</td>
      <td>6:32630797</td>
      <td>4.040244</td>
      <td>0.000053</td>
      <td>71639</td>
      <td>71639.0</td>
      <td>?+?+</td>
      <td>0.2435</td>
      <td>0.024866</td>
      <td>0.006155</td>
    </tr>
    <tr>
      <th>2</th>
      <td>6:32630824_T_C</td>
      <td>chr6</td>
      <td>32630824</td>
      <td>T</td>
      <td>C</td>
      <td>6:32630824</td>
      <td>3.921736</td>
      <td>0.000088</td>
      <td>71639</td>
      <td>71639.0</td>
      <td>?+?+</td>
      <td>0.1859</td>
      <td>0.026630</td>
      <td>0.006790</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</section>
<section id="predict-variant-effects">
<h2>Predict variant effects<a class="headerlink" href="#predict-variant-effects" title="Link to this heading">#</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">grelu.variant</span></code> module contains several functions related to analysis of variants. The <code class="docutils literal notranslate"><span class="pre">predict_variant_effects</span></code> function takes a model and a dataframe of variants, and uses the model to predict the activity of the genomic regions containing both the ref and alt alleles. It can then compare the two predictions and return an effect size for each variant. We can also apply data augmentation, i.e. make predictions for several versions of the sequence and average them together.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">grelu.variant</span>

<span class="n">odds</span> <span class="o">=</span> <span class="n">grelu</span><span class="o">.</span><span class="n">variant</span><span class="o">.</span><span class="n">predict_variant_effects</span><span class="p">(</span>
    <span class="n">variants</span><span class="o">=</span><span class="n">variants</span><span class="p">,</span>
    <span class="n">model</span><span class="o">=</span><span class="n">model</span><span class="p">,</span> 
    <span class="n">devices</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="c1"># Run on GPU 0</span>
    <span class="n">num_workers</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span>
    <span class="n">batch_size</span><span class="o">=</span><span class="mi">512</span><span class="p">,</span>
    <span class="n">genome</span><span class="o">=</span><span class="s2">&quot;hg19&quot;</span><span class="p">,</span>
    <span class="n">compare_func</span><span class="o">=</span><span class="s2">&quot;log2FC&quot;</span><span class="p">,</span> <span class="c1"># Return the log2 fold change between alt and ref predictions</span>
    <span class="n">return_ad</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="c1"># Return an anndata object.</span>
    <span class="n">rc</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="c1"># Reverse complement the ref/alt predictions and average them.</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>making dataset
Predicting DataLoader 0: 100%|█████████████████████████████████████████████████| 8/8 [00:01&lt;00:00,  7.64it/s]
</pre></div>
</div>
</div>
</div>
<p>Now this (odds) is an AnnData object (similar to <code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code> in R but in Python). Might be a good point to pause here and learn more about it: <a class="reference external" href="https://anndata.readthedocs.io/en/latest/">https://anndata.readthedocs.io/en/latest/</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">odds</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 988 × 204
    obs: &#39;snpid&#39;, &#39;chrom&#39;, &#39;pos&#39;, &#39;alt&#39;, &#39;ref&#39;, &#39;rsid&#39;, &#39;zscore&#39;, &#39;pval&#39;, &#39;nsum&#39;, &#39;neff&#39;, &#39;direction&#39;, &#39;eaf&#39;, &#39;beta&#39;, &#39;se&#39;
    var: &#39;cell type&#39;
</pre></div>
</div>
</div>
</div>
<p>odds.var contains the cell types for which the model makes predictions.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">odds</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cell type</th>
    </tr>
    <tr>
      <th>name</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Follicular</th>
      <td>Follicular</td>
    </tr>
    <tr>
      <th>Fibro General</th>
      <td>Fibro General</td>
    </tr>
    <tr>
      <th>Acinar</th>
      <td>Acinar</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>odds.obs contains the variants.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">odds</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>snpid</th>
      <th>chrom</th>
      <th>pos</th>
      <th>alt</th>
      <th>ref</th>
      <th>rsid</th>
      <th>zscore</th>
      <th>pval</th>
      <th>nsum</th>
      <th>neff</th>
      <th>direction</th>
      <th>eaf</th>
      <th>beta</th>
      <th>se</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>6:32630634_G_A</td>
      <td>chr6</td>
      <td>32630634</td>
      <td>G</td>
      <td>A</td>
      <td>6:32630634</td>
      <td>3.974476</td>
      <td>0.000071</td>
      <td>71639</td>
      <td>71639.0</td>
      <td>?+?+</td>
      <td>0.2237</td>
      <td>0.025194</td>
      <td>0.006339</td>
    </tr>
    <tr>
      <th>1</th>
      <td>6:32630797_A_G</td>
      <td>chr6</td>
      <td>32630797</td>
      <td>A</td>
      <td>G</td>
      <td>6:32630797</td>
      <td>4.040244</td>
      <td>0.000053</td>
      <td>71639</td>
      <td>71639.0</td>
      <td>?+?+</td>
      <td>0.2435</td>
      <td>0.024866</td>
      <td>0.006155</td>
    </tr>
    <tr>
      <th>2</th>
      <td>6:32630824_T_C</td>
      <td>chr6</td>
      <td>32630824</td>
      <td>T</td>
      <td>C</td>
      <td>6:32630824</td>
      <td>3.921736</td>
      <td>0.000088</td>
      <td>71639</td>
      <td>71639.0</td>
      <td>?+?+</td>
      <td>0.1859</td>
      <td>0.026630</td>
      <td>0.006790</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>And odds.X contains the predicted effect size for each variant in a numpy array of shape (n_variants x n_cell types)</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">odds</span><span class="o">.</span><span class="n">X</span><span class="p">[:</span><span class="mi">5</span><span class="p">,</span> <span class="p">:</span><span class="mi">5</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[[ 1.1112168e+00  1.1200457e+00  1.3333474e+00  1.7155571e+00
   1.6624324e+00]
 [ 2.9757369e-02 -7.0971392e-02 -6.8057328e-04  1.2972704e-01
   6.9289096e-02]
 [-5.2687943e-01 -7.6652773e-02 -7.0005655e-01 -5.5351615e-01
  -6.9800067e-01]
 [ 5.9476948e-01  4.8713604e-01  7.1450031e-01  2.2411114e-01
   4.1950190e-01]
 [-1.2315573e+00 -8.5888767e-01 -4.4588929e-01  3.2726651e-01
   1.3142172e-01]]
</pre></div>
</div>
</div>
</div>
</section>
<section id="select-a-variant-with-effect-specific-to-microglia">
<h2>Select a variant with effect specific to microglia<a class="headerlink" href="#select-a-variant-with-effect-specific-to-microglia" title="Link to this heading">#</a></h2>
<p>As an example, we search for variants that strongly disrupt accessibility in microglia but not in all cell types.</p>
<p>We set some arbitrary thresholds; The log2 fold change of the alt allele w.r.t the ref allele should be &lt; -2, and the average log2 fold change of the alt allele w.r.t the ref allele should be nonnegative.</p>
<p>This implies that the alt allele should decrease accessibility in microglia to &lt; 1/4 of its reference value, but should not decrease accessibility across all cell types on average.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Calculate the mean variant effect across all cell types</span>
<span class="n">mean_variant_effect</span> <span class="o">=</span> <span class="n">odds</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span> 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Calculate the variant effect only in microglia</span>
<span class="n">microglia_effect</span> <span class="o">=</span> <span class="n">odds</span><span class="p">[:,</span> <span class="s1">&#39;Microglia&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">squeeze</span><span class="p">()</span> 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">specific_variant_idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span>
    <span class="p">(</span><span class="n">microglia_effect</span> <span class="o">&lt;</span> <span class="o">-</span><span class="mi">2</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">mean_variant_effect</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">)</span> 
<span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">specific_variant_idx</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([498])
</pre></div>
</div>
</div>
</div>
<p>We will isolate this microglia specific variant to analyze further.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">variant</span> <span class="o">=</span> <span class="n">odds</span><span class="p">[</span><span class="n">specific_variant_idx</span><span class="p">,</span> <span class="p">:]</span> 
<span class="n">variant</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>View of AnnData object with n_obs × n_vars = 1 × 204
    obs: &#39;snpid&#39;, &#39;chrom&#39;, &#39;pos&#39;, &#39;alt&#39;, &#39;ref&#39;, &#39;rsid&#39;, &#39;zscore&#39;, &#39;pval&#39;, &#39;nsum&#39;, &#39;neff&#39;, &#39;direction&#39;, &#39;eaf&#39;, &#39;beta&#39;, &#39;se&#39;
    var: &#39;cell type&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">variant</span><span class="o">.</span><span class="n">obs</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>snpid</th>
      <th>chrom</th>
      <th>pos</th>
      <th>alt</th>
      <th>ref</th>
      <th>rsid</th>
      <th>zscore</th>
      <th>pval</th>
      <th>nsum</th>
      <th>neff</th>
      <th>direction</th>
      <th>eaf</th>
      <th>beta</th>
      <th>se</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>498</th>
      <td>6:76659840_T_G</td>
      <td>chr6</td>
      <td>76659840</td>
      <td>T</td>
      <td>G</td>
      <td>rs138940043</td>
      <td>4.009219</td>
      <td>0.000061</td>
      <td>363637</td>
      <td>363637.0</td>
      <td>??+?</td>
      <td>0.00656</td>
      <td>0.058235</td>
      <td>0.014525</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Let’s look at the effect size (log2FC) for this variant:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">variant_effect_size</span> <span class="o">=</span> <span class="n">variant</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;Microglia&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">X</span>
<span class="n">variant_effect_size</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>ArrayView([[-2.7398472]], dtype=float32)
</pre></div>
</div>
</div>
</div>
</section>
<section id="view-importance-scores-in-microglia-for-the-bases-surrounding-the-selected-variant">
<h2>View importance scores in microglia for the bases surrounding the selected variant<a class="headerlink" href="#view-importance-scores-in-microglia-for-the-bases-surrounding-the-selected-variant" title="Link to this heading">#</a></h2>
<p>We can use several different approaches to calculate the importance of each base in the reference and alternate sequence. However, we want to score each base specifically by its importance to the model’s prediction in microglia and not the model’s predictions on other tasks.</p>
<p>To do this we define a transform that operates on the model’s predictions and selects only the predictions in Microglia.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">grelu.transforms.prediction_transforms</span><span class="w"> </span><span class="kn">import</span> <span class="n">Aggregate</span>
<span class="n">microglia_score</span> <span class="o">=</span> <span class="n">Aggregate</span><span class="p">(</span><span class="n">tasks</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Microglia&quot;</span><span class="p">],</span> <span class="n">model</span><span class="o">=</span><span class="n">model</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Next, we extract the 200 bp long genomic sequences containing the reference and alternate alleles.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ref_seq</span><span class="p">,</span> <span class="n">alt_seq</span> <span class="o">=</span> <span class="n">grelu</span><span class="o">.</span><span class="n">variant</span><span class="o">.</span><span class="n">variant_to_seqs</span><span class="p">(</span>
    <span class="n">seq_len</span><span class="o">=</span><span class="n">model</span><span class="o">.</span><span class="n">data_params</span><span class="p">[</span><span class="s1">&#39;train&#39;</span><span class="p">][</span><span class="s1">&#39;seq_len&#39;</span><span class="p">],</span> <span class="c1"># Create variant-centered sequence of this length</span>
    <span class="n">genome</span><span class="o">=</span><span class="s1">&#39;hg19&#39;</span><span class="p">,</span>
    <span class="o">**</span><span class="n">variant</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">][[</span><span class="s2">&quot;chrom&quot;</span><span class="p">,</span> <span class="s2">&quot;pos&quot;</span><span class="p">,</span> <span class="s2">&quot;ref&quot;</span><span class="p">,</span> <span class="s2">&quot;alt&quot;</span><span class="p">]]</span>
<span class="p">)</span>

<span class="n">ref_seq</span><span class="p">,</span> <span class="n">alt_seq</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(&#39;CAAAGATATATGTCATAATAATTATCTTTTCACTTTCTTTGTTAGGGACCAGGAATGATAAACCACTTAGTCATTTTTTAGGTTTACAAGAACTTAAGGGGAACTAAGAAAGGAACCCTTACTCCTGAACTCTCAGCCTCATCTGTGCTGGACCATTCTAACTTTGTACCCTTTCATGAGATTGATATAATTTAGAAAAT&#39;,
 &#39;CAAAGATATATGTCATAATAATTATCTTTTCACTTTCTTTGTTAGGGACCAGGAATGATAAACCACTTAGTCATTTTTTAGGTTTACAAGAACTTAAGGTGAACTAAGAAAGGAACCCTTACTCCTGAACTCTCAGCCTCATCTGTGCTGGACCATTCTAACTTTGTACCCTTTCATGAGATTGATATAATTTAGAAAAT&#39;)
</pre></div>
</div>
</div>
</div>
<p>We are now ready to calculate the per-base importance scores. For this, we use the <code class="docutils literal notranslate"><span class="pre">grelu.interpret.score</span></code> module, which contains functions related to scoring the importance of each base in a sequence. We choose the “inputxgradient” method to score each base.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">grelu.interpret.score</span>

<span class="n">ref_attrs</span> <span class="o">=</span> <span class="n">grelu</span><span class="o">.</span><span class="n">interpret</span><span class="o">.</span><span class="n">score</span><span class="o">.</span><span class="n">get_attributions</span><span class="p">(</span>
    <span class="n">model</span><span class="p">,</span> <span class="n">ref_seq</span><span class="p">,</span> <span class="n">prediction_transform</span><span class="o">=</span><span class="n">microglia_score</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
    <span class="n">seed</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;inputxgradient&quot;</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">alt_attrs</span> <span class="o">=</span> <span class="n">grelu</span><span class="o">.</span><span class="n">interpret</span><span class="o">.</span><span class="n">score</span><span class="o">.</span><span class="n">get_attributions</span><span class="p">(</span>
    <span class="n">model</span><span class="p">,</span> <span class="n">alt_seq</span><span class="p">,</span> <span class="n">prediction_transform</span><span class="o">=</span><span class="n">microglia_score</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
    <span class="n">seed</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;inputxgradient&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
<p>We can visualize the attribution scores for the sequence, highlighting the mutated base. Here, we visualize the attributions for the central 100 bp of the 200 bp sequence.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">grelu.visualize</span>
<span class="n">grelu</span><span class="o">.</span><span class="n">visualize</span><span class="o">.</span><span class="n">plot_attributions</span><span class="p">(</span>
    <span class="n">ref_attrs</span><span class="p">,</span> <span class="n">start_pos</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">end_pos</span><span class="o">=</span><span class="mi">150</span><span class="p">,</span>
    <span class="n">highlight_positions</span><span class="o">=</span><span class="p">[</span><span class="mi">99</span><span class="p">],</span> <span class="n">ticks</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
    <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;red&#39;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;logomaker.src.Logo.Logo at 0x7f4173ba3b10&gt;
</pre></div>
</div>
<img alt="../_images/5491aab556d3fdd871a8baa671c9fab42dfb35173a9217bb9e38b09f9786e0e5.png" src="../_images/5491aab556d3fdd871a8baa671c9fab42dfb35173a9217bb9e38b09f9786e0e5.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">grelu</span><span class="o">.</span><span class="n">visualize</span><span class="o">.</span><span class="n">plot_attributions</span><span class="p">(</span>
    <span class="n">alt_attrs</span><span class="p">,</span> <span class="n">start_pos</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">end_pos</span><span class="o">=</span><span class="mi">150</span><span class="p">,</span>
    <span class="n">highlight_positions</span><span class="o">=</span><span class="p">[</span><span class="mi">99</span><span class="p">],</span> <span class="n">ticks</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
    <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;logomaker.src.Logo.Logo at 0x7f4173868550&gt;
</pre></div>
</div>
<img alt="../_images/c91f0122f80e3cf99c70a5a1f4bc5d417dead1d5f9e6f69eca6b044067155eea.png" src="../_images/c91f0122f80e3cf99c70a5a1f4bc5d417dead1d5f9e6f69eca6b044067155eea.png" />
</div>
</div>
</section>
<section id="ism">
<h2>ISM<a class="headerlink" href="#ism" title="Link to this heading">#</a></h2>
<p>We can also perform ISM (In silico mutagenesis) of the bases surrounding the variant to see what the effect would be if we mutated the reference allele to any other base. The <code class="docutils literal notranslate"><span class="pre">ISM_predict</span></code> function in <code class="docutils literal notranslate"><span class="pre">grelu.interpret.score</span></code> performs every possible single-base substitution on the given sequence, predicts the effect of each substitution, and optionally compares these predictions to the reference sequence to return an effect size for each substitution.</p>
<p>Once again, since we are interested in how important each base is to the model’s prediction in microglia,
we use the <code class="docutils literal notranslate"><span class="pre">microglia_score</span></code> transform. This ensures that ISM will score each base’s importance to the
microglial prediction only.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ism</span> <span class="o">=</span> <span class="n">grelu</span><span class="o">.</span><span class="n">interpret</span><span class="o">.</span><span class="n">score</span><span class="o">.</span><span class="n">ISM_predict</span><span class="p">(</span>
    <span class="n">ref_seq</span><span class="p">,</span>
    <span class="n">model</span><span class="p">,</span>
    <span class="n">prediction_transform</span><span class="o">=</span><span class="n">microglia_score</span><span class="p">,</span> <span class="c1"># Focus on the prediction in microglia</span>
    <span class="n">compare_func</span> <span class="o">=</span> <span class="s2">&quot;log2FC&quot;</span><span class="p">,</span> <span class="c1"># Return the log2FC of the mutated sequence prediction w.r.t the reference sequence</span>
    <span class="n">devices</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="c1"># Index of the GPU to use</span>
    <span class="n">num_workers</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Predicting DataLoader 0: 100%|███████████████████████████████████████████████| 13/13 [00:00&lt;00:00, 48.07it/s]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ism</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>C</th>
      <th>A</th>
      <th>A</th>
      <th>A</th>
      <th>G</th>
      <th>A</th>
      <th>T</th>
      <th>A</th>
      <th>T</th>
      <th>A</th>
      <th>...</th>
      <th>T</th>
      <th>T</th>
      <th>T</th>
      <th>A</th>
      <th>G</th>
      <th>A</th>
      <th>A</th>
      <th>A</th>
      <th>A</th>
      <th>T</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>A</th>
      <td>0.058220</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>-0.002522</td>
      <td>0.000000</td>
      <td>-0.061980</td>
      <td>0.000000</td>
      <td>0.039374</td>
      <td>0.000000</td>
      <td>...</td>
      <td>0.022865</td>
      <td>-0.201576</td>
      <td>-0.112543</td>
      <td>0.001650</td>
      <td>-0.101725</td>
      <td>0.001650</td>
      <td>0.001650</td>
      <td>0.001650</td>
      <td>0.001650</td>
      <td>0.037523</td>
    </tr>
    <tr>
      <th>C</th>
      <td>0.000000</td>
      <td>-0.025550</td>
      <td>-0.000935</td>
      <td>-0.064350</td>
      <td>-0.059129</td>
      <td>-0.027579</td>
      <td>-0.078199</td>
      <td>-0.031715</td>
      <td>-0.049933</td>
      <td>-0.091570</td>
      <td>...</td>
      <td>0.178785</td>
      <td>-0.174459</td>
      <td>0.084045</td>
      <td>0.150017</td>
      <td>-0.071348</td>
      <td>0.215759</td>
      <td>0.111641</td>
      <td>0.128501</td>
      <td>0.096332</td>
      <td>0.112648</td>
    </tr>
    <tr>
      <th>G</th>
      <td>0.024304</td>
      <td>0.015685</td>
      <td>-0.011021</td>
      <td>-0.106040</td>
      <td>0.000000</td>
      <td>-0.299975</td>
      <td>-0.042261</td>
      <td>-0.045126</td>
      <td>-0.057420</td>
      <td>-0.320149</td>
      <td>...</td>
      <td>0.073225</td>
      <td>0.003626</td>
      <td>0.060202</td>
      <td>0.128809</td>
      <td>0.001650</td>
      <td>0.121120</td>
      <td>0.076424</td>
      <td>0.069966</td>
      <td>0.043495</td>
      <td>0.112302</td>
    </tr>
    <tr>
      <th>T</th>
      <td>0.026044</td>
      <td>-0.029737</td>
      <td>0.037190</td>
      <td>-0.140797</td>
      <td>0.032200</td>
      <td>-0.049836</td>
      <td>0.000000</td>
      <td>-0.112758</td>
      <td>0.000000</td>
      <td>-0.053747</td>
      <td>...</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.001650</td>
      <td>-0.039917</td>
      <td>-0.161802</td>
      <td>0.090020</td>
      <td>-0.189904</td>
      <td>-0.020406</td>
      <td>-0.037939</td>
      <td>0.001650</td>
    </tr>
  </tbody>
</table>
<p>4 rows × 200 columns</p>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">grelu</span><span class="o">.</span><span class="n">visualize</span><span class="o">.</span><span class="n">plot_ISM</span><span class="p">(</span>
    <span class="n">ism</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;heatmap&quot;</span><span class="p">,</span> <span class="n">center</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mf">1.5</span><span class="p">),</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;Axes: &gt;
</pre></div>
</div>
<img alt="../_images/5c3d9d8f985de20a96050dec6b4a12772ecd0219504183010e9c4bd22c7ff4cd.png" src="../_images/5c3d9d8f985de20a96050dec6b4a12772ecd0219504183010e9c4bd22c7ff4cd.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">grelu</span><span class="o">.</span><span class="n">visualize</span><span class="o">.</span><span class="n">plot_ISM</span><span class="p">(</span>
    <span class="n">ism</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;logo&#39;</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mf">1.5</span><span class="p">),</span> <span class="n">highlight_positions</span><span class="o">=</span><span class="p">[</span><span class="mi">99</span><span class="p">],</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;red&#39;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;logomaker.src.Logo.Logo at 0x7f41766838d0&gt;
</pre></div>
</div>
<img alt="../_images/cc55d703794d13ca0a00ba9b23a7e9e7984ad4d0dd7830c03224e02e50d6536f.png" src="../_images/cc55d703794d13ca0a00ba9b23a7e9e7984ad4d0dd7830c03224e02e50d6536f.png" />
</div>
</div>
</section>
<section id="scan-variant-with-jaspar-consensus-motifs">
<h2>Scan variant with JASPAR consensus motifs<a class="headerlink" href="#scan-variant-with-jaspar-consensus-motifs" title="Link to this heading">#</a></h2>
<p>We now scan the sequences immediately around the variant to identify known TF motifs that may have been disrupted by the variant.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Select the central bases of the sequence</span>
<span class="n">central_ref_seq</span> <span class="o">=</span> <span class="n">grelu</span><span class="o">.</span><span class="n">sequence</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">resize</span><span class="p">(</span><span class="n">ref_seq</span><span class="p">,</span> <span class="mi">18</span><span class="p">)</span>
<span class="n">central_alt_seq</span> <span class="o">=</span> <span class="n">grelu</span><span class="o">.</span><span class="n">sequence</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">resize</span><span class="p">(</span><span class="n">alt_seq</span><span class="p">,</span> <span class="mi">18</span><span class="p">)</span>

<span class="n">central_ref_seq</span><span class="p">,</span> <span class="n">central_alt_seq</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(&#39;ACTTAAGGGGAACTAAGA&#39;, &#39;ACTTAAGGTGAACTAAGA&#39;)
</pre></div>
</div>
</div>
</div>
<p>We can use the <code class="docutils literal notranslate"><span class="pre">grelu.interpret.motifs</span></code> module to scan these sequences with TF motifs. Here, we use a reference set of non-redundant motifs (<a class="reference external" href="https://www.vierstra.org/resources/motif_clustering">https://www.vierstra.org/resources/motif_clustering</a>) that are provided with gReLU as <code class="docutils literal notranslate"><span class="pre">consensus</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Scan with motifs</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">grelu.interpret.motifs</span>

<span class="n">scan</span> <span class="o">=</span> <span class="n">grelu</span><span class="o">.</span><span class="n">interpret</span><span class="o">.</span><span class="n">motifs</span><span class="o">.</span><span class="n">compare_motifs</span><span class="p">(</span>
    <span class="n">ref_seq</span><span class="o">=</span><span class="n">central_ref_seq</span><span class="p">,</span>
    <span class="n">alt_seq</span><span class="o">=</span><span class="n">central_alt_seq</span><span class="p">,</span>
    <span class="n">motifs</span><span class="o">=</span><span class="s2">&quot;hocomoco_v12&quot;</span><span class="p">,</span>
    <span class="n">pthresh</span><span class="o">=</span><span class="mf">5e-4</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">scan</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>sequence</th>
      <th>motif</th>
      <th>start</th>
      <th>end</th>
      <th>strand</th>
      <th>alt</th>
      <th>ref</th>
      <th>foldChange</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>ISL2.H12CORE.0.SM.B</td>
      <td>0</td>
      <td>11</td>
      <td>+</td>
      <td>0.000000</td>
      <td>9.111901</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>MZF1.H12CORE.0.P.B</td>
      <td>3</td>
      <td>15</td>
      <td>+</td>
      <td>0.000000</td>
      <td>9.261309</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>SPI1.H12CORE.0.P.B</td>
      <td>2</td>
      <td>16</td>
      <td>+</td>
      <td>0.000000</td>
      <td>6.679440</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>SPI1.H12CORE.1.S.B</td>
      <td>4</td>
      <td>16</td>
      <td>+</td>
      <td>0.000000</td>
      <td>11.517704</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>SPIB.H12CORE.1.S.C</td>
      <td>2</td>
      <td>15</td>
      <td>+</td>
      <td>0.000000</td>
      <td>5.641884</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>5</th>
      <td>SPIB.H12CORE.2.SM.B</td>
      <td>0</td>
      <td>16</td>
      <td>+</td>
      <td>0.000000</td>
      <td>4.544859</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>6</th>
      <td>ERR1.H12CORE.0.PSM.A</td>
      <td>1</td>
      <td>12</td>
      <td>+</td>
      <td>8.008491</td>
      <td>0.000000</td>
      <td>inf</td>
    </tr>
    <tr>
      <th>7</th>
      <td>NR4A2.H12CORE.0.SM.B</td>
      <td>2</td>
      <td>12</td>
      <td>+</td>
      <td>8.064205</td>
      <td>0.000000</td>
      <td>inf</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>The ref column contains the score for the motif in the reference allele-cintaining sequence and the alt column contains the score in the alternate-allele containing sequence.</p>
<p>Note that the <code class="docutils literal notranslate"><span class="pre">SPI1.H12CORE.0.P.B</span></code> motif is lost in the alternate allele-containing sequence.</p>
</section>
<section id="compare-the-variant-impact-to-a-background-distribution">
<h2>Compare the variant impact to a background distribution<a class="headerlink" href="#compare-the-variant-impact-to-a-background-distribution" title="Link to this heading">#</a></h2>
<p>We saw that the variant has a strong effect size (log2 fold change). To place this effect size in context, we create a set of background sequences by shuffling the sequence surrounding the variant while conserving dinucleotide frequency. We then insert the reference and alternate alleles in each shuffled sequence, and compute the variant effect size again.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">test</span> <span class="o">=</span> <span class="n">grelu</span><span class="o">.</span><span class="n">variant</span><span class="o">.</span><span class="n">marginalize_variants</span><span class="p">(</span>
    <span class="n">model</span><span class="o">=</span><span class="n">model</span><span class="p">,</span>
    <span class="n">variants</span><span class="o">=</span><span class="n">variant</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span>
    <span class="n">genome</span><span class="o">=</span><span class="s2">&quot;hg19&quot;</span><span class="p">,</span>
    <span class="n">prediction_transform</span><span class="o">=</span><span class="n">microglia_score</span><span class="p">,</span>
    <span class="n">seed</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
    <span class="n">devices</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
    <span class="n">n_shuffles</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
    <span class="n">compare_func</span><span class="o">=</span><span class="s2">&quot;log2FC&quot;</span><span class="p">,</span>
    <span class="n">rc</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="c1"># Average the result over both strands</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Predicting variant effects
Predicting DataLoader 0: 100%|█████████████████████████████████████████████████| 1/1 [00:00&lt;00:00, 15.52it/s]
Predicting variant effects in background sequences
Predicting DataLoader 0: 100%|█████████████████████████████████████████████████| 4/4 [00:00&lt;00:00, 42.68it/s]
Calculating background distributions
Performing 2-sided test
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">test</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{&#39;effect_size&#39;: [-2.742067337036133],
 &#39;mean&#39;: [0.009891723282635212],
 &#39;sd&#39;: [0.3251914381980896],
 &#39;pvalue&#39;: [2.615225515668632e-17]}
</pre></div>
</div>
</div>
</div>
</section>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="4_design.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Model-guided design of regulatory elements</p>
      </div>
    </a>
    <a class="right-next"
       href="6_model_zoo.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Querying the public gReLU model zoo on Weights and Biases (wandb)</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#load-the-catlas-model">Load the CATlas model</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#view-the-model-s-metadata">View the model’s metadata</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#load-alzheimer-s-variants-from-gwas-catalog-jansen-et-al-2019-meta-analysis">Load Alzheimer’s Variants from GWAS Catalog (Jansen et al. 2019 meta-analysis)</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#filter-variants">Filter variants</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#predict-variant-effects">Predict variant effects</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#select-a-variant-with-effect-specific-to-microglia">Select a variant with effect specific to microglia</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#view-importance-scores-in-microglia-for-the-bases-surrounding-the-selected-variant">View importance scores in microglia for the bases surrounding the selected variant</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#ism">ISM</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#scan-variant-with-jaspar-consensus-motifs">Scan variant with JASPAR consensus motifs</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#compare-the-variant-impact-to-a-background-distribution">Compare the variant impact to a background distribution</a></li>
</ul>
  </nav></div>

  <div class="sidebar-secondary-item">

  
  <div class="tocsection editthispage">
    <a href="https://github.com/Genentech/gReLU/edit/main/docs/tutorials/5_variant.ipynb">
      <i class="fa-solid fa-pencil"></i>
      
      
        
          Edit on GitHub
        
      
    </a>
  </div>
</div>

  <div class="sidebar-secondary-item">
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/tutorials/5_variant.ipynb"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2023, Regulatory Genomics, BRAID, Genentech.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.2.0.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>