
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Simulation-based model interpretation with gReLU &#8212; gReLU</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=3ba690ee" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/graphviz.css?v=eafc0fe6" />
  
  <!-- So that users can add custom icons -->
  <script src="../_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="../_static/documentation_options.js?v=4e48d009"></script>
    <script src="../_static/doctools.js?v=888ff710"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <link rel="canonical" href="https://genentech.github.io/gReLU/tutorials/7_simulations.html" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="unknown" />
    <meta name="docbuild:last-update" content="Aug 26, 2025"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  
    
  

<a class="navbar-brand logo" href="../index.html">
  
  
  
  
  
  
    <p class="title logo__title">gReLU</p>
  
</a></div>
    
      <div class="navbar-item">
<div class="version-switcher__container dropdown pst-js-only">
  <button id="pst-version-switcher-button-2"
    type="button"
    class="version-switcher__button btn btn-sm dropdown-toggle"
    data-bs-toggle="dropdown"
    aria-haspopup="listbox"
    aria-controls="pst-version-switcher-list-2"
    aria-label="Version switcher list"
  >
    Choose version  <!-- this text may get changed later by javascript -->
    <span class="caret"></span>
  </button>
  <div id="pst-version-switcher-list-2"
    class="version-switcher__menu dropdown-menu list-group-flush py-0"
    role="listbox" aria-labelledby="pst-version-switcher-button-2">
    <!-- dropdown will be populated by javascript on page load -->
  </div>
</div></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../readme.html">
    Overview
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="1_inference.html">
    Tutorial: Inference
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="2_finetune.html">
    Tutorial: Fine-tuning
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="3_train.html">
    Tutorial: Training
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="4_design.html">
    Tutorial: Design
  </a>
</li>

            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button"
                data-bs-toggle="dropdown" aria-expanded="false"
                aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="5_variant.html">
    Tutorial: Variant effect prediction
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="6_model_zoo.html">
    Tutorial: Model zoo
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../contributing.html">
    Contributing
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../license.html">
    License
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../authors.html">
    Contributors
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../changelog.html">
    Changelog
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../autoapi/index.html">
    API Reference
  </a>
</li>

                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/Genentech/gReLU" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fab fa-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar hide-on-wide">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../readme.html">
    Overview
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="1_inference.html">
    Tutorial: Inference
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="2_finetune.html">
    Tutorial: Fine-tuning
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="3_train.html">
    Tutorial: Training
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="4_design.html">
    Tutorial: Design
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="5_variant.html">
    Tutorial: Variant effect prediction
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="6_model_zoo.html">
    Tutorial: Model zoo
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../contributing.html">
    Contributing
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../license.html">
    License
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../authors.html">
    Contributors
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../changelog.html">
    Changelog
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../autoapi/index.html">
    API Reference
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/Genentech/gReLU" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fab fa-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">Simulation-based model interpretation with gReLU</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="simulation-based-model-interpretation-with-grelu">
<h1>Simulation-based model interpretation with gReLU<a class="headerlink" href="#simulation-based-model-interpretation-with-grelu" title="Link to this heading">#</a></h1>
<p>This tutorial demonstrates some of gReLU’s functions for making predictions on simulated genomic sequences with specific properties. This allows us to decipher the regulatory grammar and syntax learned by a model.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scipy.stats</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">plotnine</span><span class="w"> </span><span class="kn">import</span> <span class="o">*</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">seaborn</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sns</span>

<span class="o">%</span><span class="k">matplotlib</span> inline
<span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="s2">&quot;WANDB_SILENT&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;true&quot;</span>
</pre></div>
</div>
</div>
</div>
<section id="part-1-catlas-atac-seq-binary-classification-model">
<h2>Part 1: CATlas ATAC-seq binary classification model<a class="headerlink" href="#part-1-catlas-atac-seq-binary-classification-model" title="Link to this heading">#</a></h2>
<p>First we demonstrate some simple functions using the CATLAS binary classification model. We begin by loading the model from the model zoo:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">grelu.resources</span>

<span class="n">catlas</span> <span class="o">=</span> <span class="n">grelu</span><span class="o">.</span><span class="n">resources</span><span class="o">.</span><span class="n">load_model</span><span class="p">(</span>
    <span class="n">project</span><span class="o">=</span><span class="s2">&quot;human-atac-catlas&quot;</span><span class="p">,</span>
    <span class="n">model_name</span><span class="o">=</span><span class="s2">&quot;model&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
<p>This object contains the model metadata, so we can extract a dataframe describing the cell types for which the model makes predictions.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tasks</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">catlas</span><span class="o">.</span><span class="n">data_params</span><span class="p">[</span><span class="s1">&#39;tasks&#39;</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">tasks</span><span class="p">))</span>
<span class="n">tasks</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>204
</pre></div>
</div>
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>cell type</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Follicular</td>
      <td>Follicular</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Fibro General</td>
      <td>Fibro General</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Acinar</td>
      <td>Acinar</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>We can also extract genomic intervals that were in the test set of this model:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">test_intervals</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">catlas</span><span class="o">.</span><span class="n">data_params</span><span class="p">[</span><span class="s1">&#39;test&#39;</span><span class="p">][</span><span class="s1">&#39;intervals&#39;</span><span class="p">])</span>
<span class="n">test_intervals</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>chrom</th>
      <th>start</th>
      <th>end</th>
      <th>cre_class</th>
      <th>in_fetal</th>
      <th>in_adult</th>
      <th>cre_module</th>
      <th>width</th>
      <th>cre_idx</th>
      <th>enformer_split</th>
      <th>split</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr1</td>
      <td>143497510</td>
      <td>143497710</td>
      <td>Promoter Proximal</td>
      <td>no</td>
      <td>yes</td>
      <td>113</td>
      <td>400</td>
      <td>53530</td>
      <td>test</td>
      <td>test</td>
    </tr>
    <tr>
      <th>1</th>
      <td>chr1</td>
      <td>143498052</td>
      <td>143498252</td>
      <td>Promoter Proximal</td>
      <td>no</td>
      <td>yes</td>
      <td>4</td>
      <td>400</td>
      <td>53531</td>
      <td>test</td>
      <td>test</td>
    </tr>
    <tr>
      <th>2</th>
      <td>chr1</td>
      <td>143498633</td>
      <td>143498833</td>
      <td>Promoter</td>
      <td>yes</td>
      <td>no</td>
      <td>46</td>
      <td>400</td>
      <td>53532</td>
      <td>test</td>
      <td>test</td>
    </tr>
    <tr>
      <th>3</th>
      <td>chr1</td>
      <td>143563536</td>
      <td>143563736</td>
      <td>Distal</td>
      <td>yes</td>
      <td>no</td>
      <td>63</td>
      <td>400</td>
      <td>53533</td>
      <td>test</td>
      <td>test</td>
    </tr>
    <tr>
      <th>4</th>
      <td>chr1</td>
      <td>143575231</td>
      <td>143575431</td>
      <td>Distal</td>
      <td>no</td>
      <td>yes</td>
      <td>28</td>
      <td>400</td>
      <td>53534</td>
      <td>test</td>
      <td>test</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Let us extract 100 intervals from the test set:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">test_intervals</span> <span class="o">=</span> <span class="n">test_intervals</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<section id="motif-marginalization">
<h3>1.1 Motif marginalization<a class="headerlink" href="#motif-marginalization" title="Link to this heading">#</a></h3>
<p>In order to understand the function of a specific motif, we can insert the motif into shuffled sequences. By making predictions on the shuffled sequences before and after motif insertion, we can compute the extent to which the motif alone increases predicted activity.</p>
<p>To demonstrate, we extract a few selected motifs from the HOCOMOCO database.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">grelu.io.motifs</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">grelu.interpret.motifs</span>

<span class="n">motif_names</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CTCF.H12CORE.0.P.B&#39;</span><span class="p">,</span> <span class="s1">&#39;IRF4.H12CORE.0.P.B&#39;</span><span class="p">,</span> <span class="s1">&#39;SPI1.H12CORE.0.P.B&#39;</span><span class="p">,</span>
               <span class="s1">&#39;HNF1A.H12CORE.0.PS.A&#39;</span><span class="p">]</span>
<span class="n">motifs</span> <span class="o">=</span> <span class="n">grelu</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">motifs</span><span class="o">.</span><span class="n">read_meme_file</span><span class="p">(</span><span class="s2">&quot;hocomoco_v12&quot;</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="n">motif_names</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>This is a dictionary containing the motifs. We can extract the consensus sequence for each motif:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">patterns</span> <span class="o">=</span> <span class="n">grelu</span><span class="o">.</span><span class="n">interpret</span><span class="o">.</span><span class="n">motifs</span><span class="o">.</span><span class="n">motifs_to_strings</span><span class="p">(</span><span class="n">motifs</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">patterns</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[&#39;TGGCCACTAGGGGGCGCTAG&#39;, &#39;AAAGAGGAACTGAAACT&#39;, &#39;AAAAGAGGAAGTGA&#39;, &#39;GGTTAATGATTAACC&#39;]
</pre></div>
</div>
</div>
</div>
<p>Now, we insert each of these motifs into shuffled sequences and compare the predictions before and after motif insertion. gReLU provides an easy function <code class="docutils literal notranslate"><span class="pre">marginalize_patterns</span></code> to perform this analysis.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">grelu.interpret.simulate</span>

<span class="n">preds</span> <span class="o">=</span> <span class="n">grelu</span><span class="o">.</span><span class="n">interpret</span><span class="o">.</span><span class="n">simulate</span><span class="o">.</span><span class="n">marginalize_patterns</span><span class="p">(</span>
    <span class="n">model</span><span class="o">=</span><span class="n">catlas</span><span class="p">,</span>
    <span class="n">patterns</span><span class="o">=</span><span class="n">patterns</span><span class="p">,</span>
    <span class="n">seqs</span><span class="o">=</span><span class="n">test_intervals</span><span class="p">,</span>
    <span class="n">genome</span><span class="o">=</span><span class="s1">&#39;hg38&#39;</span><span class="p">,</span>
    <span class="n">devices</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
    <span class="n">num_workers</span><span class="o">=</span><span class="mi">16</span><span class="p">,</span>
    <span class="n">batch_size</span><span class="o">=</span><span class="mi">512</span><span class="p">,</span>
    <span class="n">n_shuffles</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="c1"># Each test sequence will be shuffled 3 times </span>
    <span class="n">seed</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
    <span class="n">compare_func</span><span class="o">=</span><span class="s1">&#39;subtract&#39;</span><span class="p">,</span> <span class="c1"># Return the difference in prediction with and without each motif. If not supplied, predictions with and without are returned separately.</span>
<span class="p">)</span><span class="o">.</span><span class="n">squeeze</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Predicting DataLoader 0: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 3/3 [00:00&lt;00:00,  3.40it/s]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">preds</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(100, 3, 4, 204)
</pre></div>
</div>
</div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">preds</span></code> object has the shape (100, 3, 4, 204) where 100 is the number of background sequences, 3 is the number of times each sequence is shuffled, 4 is the number of motifs, and 204 is the number of cell types. Let’s reshape this so that each motif has 300 background sequences:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">preds</span> <span class="o">=</span> <span class="n">preds</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="mi">300</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">204</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>We can now subset the results for our cell type of interest. For example, microglia:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">microglia_preds</span> <span class="o">=</span> <span class="n">preds</span><span class="p">[:,</span> <span class="p">:,</span> <span class="n">tasks</span><span class="o">.</span><span class="n">name</span><span class="o">==</span><span class="s1">&#39;Microglia&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">squeeze</span><span class="p">()</span>
<span class="n">microglia_preds</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(300, 4)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">microglia_preds</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">microglia_preds</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">motif_names</span><span class="p">)</span>
<span class="n">microglia_preds</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>CTCF.H12CORE.0.P.B</th>
      <th>IRF4.H12CORE.0.P.B</th>
      <th>SPI1.H12CORE.0.P.B</th>
      <th>HNF1A.H12CORE.0.PS.A</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0.028903</td>
      <td>0.602199</td>
      <td>0.533135</td>
      <td>-0.021193</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0.014234</td>
      <td>0.054343</td>
      <td>0.016284</td>
      <td>-0.002339</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0.047415</td>
      <td>0.107500</td>
      <td>0.008276</td>
      <td>-0.001469</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0.015104</td>
      <td>0.175864</td>
      <td>0.167947</td>
      <td>-0.013539</td>
    </tr>
    <tr>
      <th>4</th>
      <td>-0.053547</td>
      <td>0.719626</td>
      <td>0.607563</td>
      <td>-0.095212</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>We can now visualize the effect of each of these motifs in microglia:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">microglia_preds</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">box</span><span class="p">(</span><span class="n">rot</span><span class="o">=</span><span class="mi">45</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span><span class="mi">2</span><span class="p">),</span>
    <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Increase in predicted probability of</span><span class="se">\n</span><span class="s1">accessibility in microglia after inserting motif&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;Axes: title={&#39;center&#39;: &#39;Increase in predicted probability of\naccessibility in microglia after inserting motif&#39;}&gt;
</pre></div>
</div>
<img alt="../_images/6b57aebac651d5825d6e11ca497a6898102d14779898cdd75b0d1fe69ed7e009.png" src="../_images/6b57aebac651d5825d6e11ca497a6898102d14779898cdd75b0d1fe69ed7e009.png" />
</div>
</div>
<p>We can compute the significance of the observed changes using Wilcoxon’s signed-rank test:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scipy</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">wilcoxon</span><span class="p">(</span><span class="n">microglia_preds</span><span class="p">[</span><span class="s1">&#39;SPI1.H12CORE.0.P.B&#39;</span><span class="p">],</span> <span class="n">alternative</span><span class="o">=</span><span class="s1">&#39;greater&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>WilcoxonResult(statistic=45138.0, pvalue=3.4306111433259653e-51)
</pre></div>
</div>
</div>
</div>
<p>We can also compare the average effect of each motifs across all 204 cell types. For this, we first average the effect across all 500 peaks:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">preds</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="n">mean_motif_effect</span> <span class="o">=</span> <span class="n">preds</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">mean_motif_effect</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(300, 4, 204)
(4, 204)
</pre></div>
</div>
</div>
</div>
<p>This gives us an array of size (motifs x cell types). We can represent this as a pandas dataframe:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mean_motif_effect</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">mean_motif_effect</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">motif_names</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">tasks</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
<span class="n">mean_motif_effect</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:</span><span class="mi">2</span><span class="p">,</span> <span class="p">:</span><span class="mi">5</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>name</th>
      <th>Follicular</th>
      <th>Fibro General</th>
      <th>Acinar</th>
      <th>T Lymphocyte 1 (CD8+)</th>
      <th>T lymphocyte 2 (CD4+)</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>CTCF.H12CORE.0.P.B</th>
      <td>0.388247</td>
      <td>0.394344</td>
      <td>0.432518</td>
      <td>0.305757</td>
      <td>0.294902</td>
    </tr>
    <tr>
      <th>IRF4.H12CORE.0.P.B</th>
      <td>-0.036361</td>
      <td>-0.012824</td>
      <td>0.002282</td>
      <td>0.056297</td>
      <td>0.036171</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>For visualization, let us select the 20 cell types with highest variance:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">variable_cell_types</span> <span class="o">=</span> <span class="n">mean_motif_effect</span><span class="o">.</span><span class="n">var</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">sort_values</span><span class="p">()</span><span class="o">.</span><span class="n">tail</span><span class="p">(</span><span class="mi">20</span><span class="p">)</span><span class="o">.</span><span class="n">index</span>
<span class="n">variable_cell_types</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Index([&#39;Fetal Photoreceptor&#39;, &#39;Fetal T Lymphocyte 1 (CD4+)&#39;,
       &#39;Fetal Enterocyte 3&#39;, &#39;Fetal Enterocyte 1&#39;,
       &#39;Fetal Retinal Progenitor 1&#39;, &#39;Memory B&#39;, &#39;Fetal Acinar 2&#39;,
       &#39;Macrophage Gen or Alv&#39;, &#39;Hepatocyte&#39;, &#39;Plasma B&#39;,
       &#39;Fetal Adrenal Neuron&#39;, &#39;Alpha 1&#39;, &#39;Fetal Adrenal Cortical&#39;, &#39;Acinar&#39;,
       &#39;Fetal Thymocyte&#39;, &#39;Macrophage General&#39;,
       &#39;Fetal Syncitio+Cytotrophoblast&#39;, &#39;Fetal B Lymphocyte 1 (SPIB+)&#39;,
       &#39;Fetal Erythroblast 2&#39;, &#39;Fetal Hepatoblast&#39;],
      dtype=&#39;object&#39;, name=&#39;name&#39;)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sns</span><span class="o">.</span><span class="n">clustermap</span><span class="p">(</span><span class="n">mean_motif_effect</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">variable_cell_types</span><span class="p">]</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span><span class="mi">6</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;seaborn.matrix.ClusterGrid at 0x7ff540b3c090&gt;
</pre></div>
</div>
<img alt="../_images/9824c9287596d46e6a20ea382114bad4a3cad19d1c1b940c3e746abf67f2c26c.png" src="../_images/9824c9287596d46e6a20ea382114bad4a3cad19d1c1b940c3e746abf67f2c26c.png" />
</div>
</div>
</section>
<section id="motif-spacing">
<h3>1.2 Motif spacing<a class="headerlink" href="#motif-spacing" title="Link to this heading">#</a></h3>
<p>Given two motifs, we may want to know whether the spacing between them is important for regulatory function. Here, we insert a TEAD motif in the center of the shuffled peak and then insert a FOS motif at different distances from the TEAD motif. We compute the increase in predictions in fibroblasts before and after inserting the TEAD motif. We marginalize the effect of the FOS motif across all 500 shuffled peak sequences.</p>
<p>First, we load the motifs from the HOCOMOCO database:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">motifs</span> <span class="o">=</span> <span class="n">grelu</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">motifs</span><span class="o">.</span><span class="n">read_meme_file</span><span class="p">(</span>
    <span class="s2">&quot;hocomoco_v12&quot;</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;TEAD1.H12CORE.0.PSM.A&#39;</span><span class="p">,</span> <span class="s1">&#39;FOS.H12CORE.0.P.B&#39;</span><span class="p">]</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>And extract the consensus sequence for each motif:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">patterns</span> <span class="o">=</span> <span class="n">grelu</span><span class="o">.</span><span class="n">interpret</span><span class="o">.</span><span class="n">motifs</span><span class="o">.</span><span class="n">motifs_to_strings</span><span class="p">(</span><span class="n">motifs</span><span class="p">)</span>
<span class="n">tead_pattern</span><span class="p">,</span> <span class="n">fos_pattern</span> <span class="o">=</span> <span class="n">patterns</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">tead_pattern</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">fos_pattern</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>CCACATTCCAGGC
GTGAGTCACCA
</pre></div>
</div>
</div>
</div>
<p>gReLU provides another handy function <code class="docutils literal notranslate"><span class="pre">marginalize_pattern_spacing</span></code> to compute the effect of different inter-motif distances. This function produces two outputs: <code class="docutils literal notranslate"><span class="pre">preds</span></code> (model predictions) and <code class="docutils literal notranslate"><span class="pre">distances</span></code> (distances at which the second motif was inserted). Note that distances are calculated between the start positions of the two motifs.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">preds</span><span class="p">,</span> <span class="n">distances</span> <span class="o">=</span> <span class="n">grelu</span><span class="o">.</span><span class="n">interpret</span><span class="o">.</span><span class="n">simulate</span><span class="o">.</span><span class="n">marginalize_pattern_spacing</span><span class="p">(</span>
    <span class="n">model</span> <span class="o">=</span> <span class="n">catlas</span><span class="p">,</span>
    <span class="n">seqs</span> <span class="o">=</span> <span class="n">test_intervals</span><span class="p">,</span>
    <span class="n">fixed_pattern</span> <span class="o">=</span> <span class="n">tead_pattern</span><span class="p">,</span> <span class="c1"># this motif will be fixed at the center of the sequence</span>
    <span class="n">moving_pattern</span> <span class="o">=</span> <span class="n">fos_pattern</span><span class="p">,</span> <span class="c1"># this motif will be inserted at different distances</span>
    <span class="n">genome</span> <span class="o">=</span> <span class="s1">&#39;hg38&#39;</span><span class="p">,</span>
    <span class="n">n_shuffles</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="c1"># Each test sequence will be shuffled once</span>
    <span class="n">seed</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
    <span class="n">devices</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
    <span class="n">num_workers</span> <span class="o">=</span> <span class="mi">16</span><span class="p">,</span>
    <span class="n">batch_size</span> <span class="o">=</span> <span class="mi">512</span><span class="p">,</span>
    <span class="n">compare_func</span> <span class="o">=</span> <span class="s1">&#39;subtract&#39;</span><span class="p">,</span> <span class="c1"># Return the difference in prediction after inserting FOS</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Predicting DataLoader 0: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 33/33 [00:01&lt;00:00, 26.87it/s]
</pre></div>
</div>
</div>
</div>
<p>Note that this produces two outputs: <code class="docutils literal notranslate"><span class="pre">preds</span></code> (the model’s predictions) and <code class="docutils literal notranslate"><span class="pre">distances</span></code> (the distance between the two motifs, corresponding to the predictions.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">preds</span> <span class="o">=</span> <span class="n">preds</span><span class="o">.</span><span class="n">squeeze</span><span class="p">()</span>
<span class="n">preds</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(100, 167, 204)
</pre></div>
</div>
</div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">preds</span></code> has shape <code class="docutils literal notranslate"><span class="pre">(number</span> <span class="pre">of</span> <span class="pre">sequences,</span> <span class="pre">number</span> <span class="pre">of</span> <span class="pre">positions,</span> <span class="pre">number</span> <span class="pre">of</span> <span class="pre">tasks)</span></code>. We are going to focus on one cell type (fibroblasts). So we subset the predictions for that cell type:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">preds</span> <span class="o">=</span> <span class="n">preds</span><span class="p">[:,</span> <span class="p">:,</span> <span class="n">tasks</span><span class="o">.</span><span class="n">name</span><span class="o">==</span><span class="s1">&#39;Fibro General&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">squeeze</span><span class="p">()</span>
<span class="n">preds</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(100, 167)
</pre></div>
</div>
</div>
</div>
<p>Let’s also take a look at <code class="docutils literal notranslate"><span class="pre">distances</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">distances</span><span class="p">[:</span><span class="mi">5</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[-93, -92, -91, -90, -89]
</pre></div>
</div>
</div>
</div>
<p>A convenient way to analyze these is to plot predicted effect of the FOS motif vs. distance from the TEAD motif.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">grelu.visualize</span>
<span class="n">grelu</span><span class="o">.</span><span class="n">visualize</span><span class="o">.</span><span class="n">plot_position_effect</span><span class="p">(</span>
    <span class="n">preds</span><span class="p">,</span> <span class="n">distances</span><span class="p">,</span>
    <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Increase in predicted probability of accessibility</span><span class="se">\n</span><span class="s2">in fibroblasts after inserting FOS motif&quot;</span><span class="p">,</span>
    <span class="n">xlab</span><span class="o">=</span><span class="s2">&quot;Distance from TEAD motif&quot;</span><span class="p">,</span>
    <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span><span class="mi">2</span><span class="p">)</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../_images/c9bf97a9ad81e52eba8566884ba66b6e5b1132b76c3d927e586837093ecaf18d.png"><img alt="../_images/c9bf97a9ad81e52eba8566884ba66b6e5b1132b76c3d927e586837093ecaf18d.png" src="../_images/c9bf97a9ad81e52eba8566884ba66b6e5b1132b76c3d927e586837093ecaf18d.png" style="width: 600px; height: 200px;" /></a>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;mean_effect&#39;</span><span class="p">:</span><span class="n">preds</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="mi">0</span><span class="p">),</span> <span class="s1">&#39;dist&#39;</span><span class="p">:</span><span class="n">distances</span><span class="p">})</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s1">&#39;mean_effect&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>mean_effect</th>
      <th>dist</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>85</th>
      <td>0.314874</td>
      <td>15</td>
    </tr>
    <tr>
      <th>86</th>
      <td>0.199096</td>
      <td>16</td>
    </tr>
    <tr>
      <th>84</th>
      <td>0.192461</td>
      <td>14</td>
    </tr>
    <tr>
      <th>87</th>
      <td>0.149526</td>
      <td>17</td>
    </tr>
    <tr>
      <th>83</th>
      <td>0.143634</td>
      <td>13</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>gReLU shows us a clear distance dependence: a distance of 15 nucleotides between the start positions of the two motifs creates a composite motif that drives higher predicted accessibility. Since the TEAD motif is 15 bp long, this means two nucleotides between the end of the TEAD motif and the start of the FOS motif (CCACATTCCAGGC _ _ GTGAGTCACCA)</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">del</span> <span class="n">catlas</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="part-2-enformer">
<h2>Part 2: Enformer<a class="headerlink" href="#part-2-enformer" title="Link to this heading">#</a></h2>
<p>Next we will examine regulatory relationships between gene TSSs and their genomic context using the Enformer model.</p>
<p>First, we import the Enformer model from the model zoo.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">grelu.resources</span>
<span class="n">enformer</span> <span class="o">=</span> <span class="n">grelu</span><span class="o">.</span><span class="n">resources</span><span class="o">.</span><span class="n">load_model</span><span class="p">(</span>
    <span class="n">project</span><span class="o">=</span><span class="s2">&quot;enformer&quot;</span><span class="p">,</span>
    <span class="n">model_name</span><span class="o">=</span><span class="s2">&quot;human&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
<p>This object includes the task metadata. For the current analyses, we will focus on  a single CAGE track in K562 cells, which is track 5111.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">k562_cage_track_idx</span><span class="o">=</span><span class="mi">5111</span>
<span class="n">tasks</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">enformer</span><span class="o">.</span><span class="n">data_params</span><span class="p">[</span><span class="s1">&#39;tasks&#39;</span><span class="p">])</span>
<span class="n">tasks</span><span class="o">.</span><span class="n">iloc</span><span class="p">[[</span><span class="n">k562_cage_track_idx</span><span class="p">]]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>file</th>
      <th>clip</th>
      <th>scale</th>
      <th>sum_stat</th>
      <th>description</th>
      <th>assay</th>
      <th>sample</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>5111</th>
      <td>CNhs12336</td>
      <td>/home/drk/tillage/datasets/human/cage/fantom/C...</td>
      <td>384</td>
      <td>1</td>
      <td>sum</td>
      <td>CAGE:chronic myelogenous leukemia cell line:K5...</td>
      <td>CAGE</td>
      <td>chronic myelogenous leukemia cell line:K562 EN...</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>We are going to analyze CAGE activity at the TSS of the ABCA8 gene. This gene is located on chromosome 17 and its TSS is at position 68955392. We define a region of 5000 bp centered on this position as the TSS region.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tss_interval</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
    <span class="s1">&#39;chrom&#39;</span><span class="p">:[</span><span class="s1">&#39;chr17&#39;</span><span class="p">],</span> <span class="s1">&#39;start&#39;</span><span class="p">:[</span><span class="mi">68955392</span><span class="o">-</span><span class="mi">2500</span><span class="p">],</span> <span class="s1">&#39;end&#39;</span><span class="p">:[</span><span class="mi">68955392</span><span class="o">+</span><span class="mi">2500</span><span class="p">]</span>
<span class="p">})</span>
<span class="nb">print</span><span class="p">(</span><span class="n">tss_interval</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>   chrom     start       end
0  chr17  68952892  68957892
</pre></div>
</div>
</div>
</div>
<p>And extract its sequence from the genome:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">grelu.sequence.format</span>
<span class="n">tss_seq</span> <span class="o">=</span> <span class="n">grelu</span><span class="o">.</span><span class="n">sequence</span><span class="o">.</span><span class="n">format</span><span class="o">.</span><span class="n">convert_input_type</span><span class="p">(</span><span class="n">tss_interval</span><span class="p">,</span> <span class="s2">&quot;strings&quot;</span><span class="p">,</span> <span class="n">genome</span><span class="o">=</span><span class="s2">&quot;hg38&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="c1"># Don&#39;t try to print this!</span>
</pre></div>
</div>
</div>
</div>
<p>We will construct an input interval for Enformer centered on this TSS position.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">grelu.sequence.utils</span>
<span class="n">input_interval</span> <span class="o">=</span> <span class="n">grelu</span><span class="o">.</span><span class="n">sequence</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">resize</span><span class="p">(</span><span class="n">tss_interval</span><span class="p">,</span> <span class="n">enformer</span><span class="o">.</span><span class="n">data_params</span><span class="p">[</span><span class="s1">&#39;train&#39;</span><span class="p">][</span><span class="s1">&#39;seq_len&#39;</span><span class="p">])</span>
<span class="n">input_interval</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>chrom</th>
      <th>start</th>
      <th>end</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>chr17</td>
      <td>68857088</td>
      <td>69053696</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Let’s examine Enformer’s predictions for this interval:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">input_seq</span> <span class="o">=</span> <span class="n">grelu</span><span class="o">.</span><span class="n">sequence</span><span class="o">.</span><span class="n">format</span><span class="o">.</span><span class="n">convert_input_type</span><span class="p">(</span><span class="n">input_interval</span><span class="p">,</span> <span class="s2">&quot;strings&quot;</span><span class="p">,</span> <span class="n">genome</span><span class="o">=</span><span class="s1">&#39;hg38&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">input_pred</span> <span class="o">=</span> <span class="n">enformer</span><span class="o">.</span><span class="n">predict_on_seqs</span><span class="p">(</span><span class="n">input_seq</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">input_pred</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(1, 5313, 896)
</pre></div>
</div>
</div>
</div>
<p>Enformer predicts 5313 output tracks, but we are only interested in the predictions for the K562 CAGE track - let’s plot those, and highlight the TSS region that we defined.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span> <span class="o">=</span> <span class="n">grelu</span><span class="o">.</span><span class="n">visualize</span><span class="o">.</span><span class="n">plot_tracks</span><span class="p">(</span>
    <span class="n">input_pred</span><span class="p">[:,</span> <span class="n">k562_cage_track_idx</span><span class="p">],</span>
    <span class="n">start_pos</span><span class="o">=</span><span class="n">input_interval</span><span class="o">.</span><span class="n">start</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">end_pos</span><span class="o">=</span><span class="n">input_interval</span><span class="o">.</span><span class="n">end</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
    <span class="n">highlight_intervals</span><span class="o">=</span><span class="n">tss_interval</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/bb4d1b5cab54a1b0e7584805834c4f1cbf4035c188a7ce4f225e8ba451219621.png" src="../_images/bb4d1b5cab54a1b0e7584805834c4f1cbf4035c188a7ce4f225e8ba451219621.png" />
</div>
</div>
<p>We see that Enformer predicts transcriptional activity at the ABCA8 TSS.</p>
<p>Instead of making predictions for the entire sequence across all tracks and then subsetting the TSS and the K562 CAGE track each time, gReLU allows us to use a transform object that will compute only the relevant predictions.</p>
<p>Here, we can create an <code class="docutils literal notranslate"><span class="pre">Aggregate</span></code> transform that sums the predicted values in the bins overlapping the TSS region, in the K562 CAGE track. This transform, when applied to Enformer’s predictions, computes the statistic we’re interested in.</p>
<p>To do this, we first need the indices of the output bins that overlap with the TSS region:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">enformer</span><span class="o">.</span><span class="n">input_intervals_to_output_bins</span><span class="p">(</span><span class="n">tss_interval</span><span class="p">,</span> <span class="n">start_pos</span><span class="o">=</span><span class="n">input_interval</span><span class="o">.</span><span class="n">start</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>start</th>
      <th>end</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>428</td>
      <td>468</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Now we can create the transform:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">grelu.transforms.prediction_transforms</span><span class="w"> </span><span class="kn">import</span> <span class="n">Aggregate</span>
<span class="n">tss_activity_transform</span> <span class="o">=</span> <span class="n">Aggregate</span><span class="p">(</span>
    <span class="n">tasks</span><span class="o">=</span><span class="n">k562_cage_track_idx</span><span class="p">,</span>
    <span class="n">positions</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">428</span><span class="p">,</span> <span class="mi">468</span><span class="p">)),</span> <span class="c1"># these are the indices of the central bins overlapping with the TSS</span>
    <span class="n">length_aggfunc</span><span class="o">=</span><span class="s1">&#39;sum&#39;</span> <span class="c1"># Sum the output over all bins</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>So we can now compute the TSS activity in the wild-type sequence:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">WT</span> <span class="o">=</span> <span class="n">tss_activity_transform</span><span class="o">.</span><span class="n">compute</span><span class="p">(</span><span class="n">input_pred</span><span class="p">)</span><span class="o">.</span><span class="n">squeeze</span><span class="p">()</span>
<span class="n">WT</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array(9.00075, dtype=float32)
</pre></div>
</div>
</div>
</div>
<p>Next, we will simulate different changes to this locus and see how this statistic changes.</p>
<section id="context-dependence-test">
<h3>2.1 Context dependence test<a class="headerlink" href="#context-dependence-test" title="Link to this heading">#</a></h3>
<p>The context dependence test introduced in CREME (<a class="reference external" href="https://doi.org/10.1101/2023.07.03.547592">https://doi.org/10.1101/2023.07.03.547592</a>) measures the extent to which a prediction of a given TSS activity is influenced by its context.</p>
<p>It includes the following steps:</p>
<ul class="simple">
<li><p>Predict TSS activity for the wild type sequence (denoted as WT).</p></li>
<li><p>Dinucleotide shuffle the sequence (except the 5kb tile centered at the TSS).</p></li>
<li><p>Predict TSS activity for the shuffled sequence (denoted as MUTANT).</p></li>
<li><p>Normalization: compute context effect on TSS using WT as control: (WT - MUTANT) / WT</p></li>
<li><p>Repeat steps 2-4 10 times and average across different random dinucleotide shuffles.</p></li>
</ul>
<p>A positive effect size means that the central TSS prediction for the mutated sequence is lower than WT, which indicates that we have perturbed an enhancing context. Negative effect size means that the central TSS prediction for the mutated sequence is higher than WT, which suggests that we have perturbed a silencing context.</p>
<p>We have already done the first step, which is to predict TSS activity in the wild-type (WT). So, we move on to the subsequent steps.</p>
<p>Using gReLU, we can shuffle the input sequence 10 times.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">shuffled_seqs</span> <span class="o">=</span> <span class="n">grelu</span><span class="o">.</span><span class="n">sequence</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">dinuc_shuffle</span><span class="p">(</span><span class="n">input_seq</span><span class="p">,</span> <span class="n">n_shuffles</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>And re-insert the TSS region into the center of the shuffled sequences:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">grelu.sequence.mutate</span><span class="w"> </span><span class="kn">import</span> <span class="n">mutate</span>
<span class="n">shuf_seqs_w_tss</span> <span class="o">=</span> <span class="p">[</span><span class="n">mutate</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">tss_seq</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">shuffled_seqs</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<p>We now make predictions for the TSS in the shuffled sequences:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">grelu.data.dataset</span><span class="w"> </span><span class="kn">import</span> <span class="n">SeqDataset</span>
<span class="n">ds</span> <span class="o">=</span> <span class="n">SeqDataset</span><span class="p">(</span><span class="n">shuf_seqs_w_tss</span><span class="p">)</span>

<span class="c1"># Make predictions</span>
<span class="n">MUTANT</span> <span class="o">=</span> <span class="n">enformer</span><span class="o">.</span><span class="n">predict_on_dataset</span><span class="p">(</span><span class="n">ds</span><span class="p">,</span> <span class="n">devices</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">batch_size</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>

<span class="c1"># Apply the transform to compute predicted TSS activity</span>
<span class="n">MUTANT</span> <span class="o">=</span> <span class="n">tss_activity_transform</span><span class="o">.</span><span class="n">compute</span><span class="p">(</span><span class="n">MUTANT</span><span class="p">)</span><span class="o">.</span><span class="n">squeeze</span><span class="p">()</span>

<span class="n">MUTANT</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Predicting DataLoader 0: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00&lt;00:00,  2.11it/s]
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([1.7110163, 1.7958962, 1.6829112, 1.7478762, 1.6674683, 1.6271539,
       1.6110293, 1.7591425, 1.614184 , 1.6504713], dtype=float32)
</pre></div>
</div>
</div>
</div>
<p>And compute the final metric:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">result</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">((</span><span class="n">WT</span> <span class="o">-</span> <span class="n">MUTANT</span><span class="p">)</span> <span class="o">/</span> <span class="n">WT</span><span class="p">)</span>
<span class="n">result</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>0.8126029
</pre></div>
</div>
</div>
</div>
<p>The positive value shows that the TSS is in an enhancing context.</p>
</section>
<section id="enhancer-discovery-by-in-silico-tiling">
<h3>2.2 Enhancer Discovery by in silico tiling<a class="headerlink" href="#enhancer-discovery-by-in-silico-tiling" title="Link to this heading">#</a></h3>
<p>Which parts of the input sequence are responsible for its enhancing context? We can discover putative regulatory elements in silico by shuffling successive tiles in the sequence and computing the change in predicted TSS activity. gReLU provides a shuffle_tiles function for this analysis:</p>
<p>Which parts of the input sequence are responsible for its enhancing context? We can discover putative regulatory elements <i>in silico</i> by shuffling successive tiles in the sequence and computing the change in predicted TSS activity. gReLU provides a <code class="docutils literal notranslate"><span class="pre">shuffle_tiles</span></code> function for this analysis:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">preds</span><span class="p">,</span> <span class="n">tiles</span> <span class="o">=</span> <span class="n">grelu</span><span class="o">.</span><span class="n">interpret</span><span class="o">.</span><span class="n">simulate</span><span class="o">.</span><span class="n">shuffle_tiles</span><span class="p">(</span>
    <span class="n">model</span><span class="o">=</span><span class="n">enformer</span><span class="p">,</span>
    <span class="n">seqs</span><span class="o">=</span><span class="p">[</span><span class="n">input_seq</span><span class="p">],</span>
    <span class="n">tile_len</span><span class="o">=</span><span class="mi">5000</span><span class="p">,</span> <span class="c1"># Length of region to shuffle</span>
    <span class="n">stride</span><span class="o">=</span><span class="mi">2500</span><span class="p">,</span> <span class="c1"># Tiles will overlap by this much</span>
    <span class="n">protect_center</span><span class="o">=</span> <span class="mi">5000</span><span class="p">,</span> <span class="c1"># Do not disrupt the central TSS region</span>
    <span class="n">n_shuffles</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="c1"># Number of times to shuffle each tile</span>
    <span class="n">seed</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
    <span class="n">devices</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
    <span class="n">num_workers</span><span class="o">=</span><span class="mi">16</span><span class="p">,</span>
    <span class="n">batch_size</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span>
    <span class="n">prediction_transform</span><span class="o">=</span><span class="n">tss_activity_transform</span><span class="p">,</span>
    <span class="n">compare_func</span><span class="o">=</span><span class="s1">&#39;subtract&#39;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Predicting DataLoader 0: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00&lt;00:00, 10.14it/s]
Predicting DataLoader 0: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 46/46 [00:25&lt;00:00,  1.80it/s]
</pre></div>
</div>
</div>
</div>
<p>Note that this produces two outputs: <code class="docutils literal notranslate"><span class="pre">preds</span></code> (the model’s predictions) and <code class="docutils literal notranslate"><span class="pre">tiles</span></code> (the regions that were shuffled, corresponding to the predictions).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">tiles</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="n">tiles</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(73, 2)
</pre></div>
</div>
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>start</th>
      <th>end</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>5000</td>
    </tr>
    <tr>
      <th>1</th>
      <td>2500</td>
      <td>7500</td>
    </tr>
    <tr>
      <th>2</th>
      <td>5000</td>
      <td>10000</td>
    </tr>
    <tr>
      <th>3</th>
      <td>7500</td>
      <td>12500</td>
    </tr>
    <tr>
      <th>4</th>
      <td>10000</td>
      <td>15000</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">preds</span> <span class="o">=</span> <span class="n">preds</span><span class="o">.</span><span class="n">squeeze</span><span class="p">()</span>
<span class="n">preds</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(73, 5)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">grelu</span><span class="o">.</span><span class="n">visualize</span><span class="o">.</span><span class="n">plot_position_effect</span><span class="p">(</span>
    <span class="n">preds</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">tiles</span><span class="o">.</span><span class="n">start</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span>
    <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Change in predicted TSS activity</span><span class="se">\n</span><span class="s2">after shuffling tile&quot;</span><span class="p">,</span>
    <span class="n">xlab</span><span class="o">=</span><span class="s2">&quot;Start position of tile&quot;</span><span class="p">,</span>
    <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span><span class="mi">2</span><span class="p">)</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../_images/936860ad117826a0ec13d68305962b56ebc3016132da6834efa9e7d34ceca122.png"><img alt="../_images/936860ad117826a0ec13d68305962b56ebc3016132da6834efa9e7d34ceca122.png" src="../_images/936860ad117826a0ec13d68305962b56ebc3016132da6834efa9e7d34ceca122.png" style="width: 688px; height: 288px;" /></a>
</div>
</div>
<p>The tiles with the strongest negative value are the putative enhancers. For a tabular representation, we can add the predictions to the dataframe, averaging across all shuffles of the same sequence, then sort the tiles by effect:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tiles</span><span class="p">[</span><span class="s1">&#39;effect&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">preds</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tiles</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s1">&#39;effect&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>start</th>
      <th>end</th>
      <th>effect</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>20</th>
      <td>50000</td>
      <td>55000</td>
      <td>-7.329542</td>
    </tr>
    <tr>
      <th>19</th>
      <td>47500</td>
      <td>52500</td>
      <td>-7.232471</td>
    </tr>
    <tr>
      <th>37</th>
      <td>102500</td>
      <td>107500</td>
      <td>-6.966073</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>It seems that the tile from position 50,000 - 55,000 of the input sequence has the strongest enhancer activity. We will define this as our putative CRE and extract its sequence.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cre_pos</span> <span class="o">=</span> <span class="mi">50000</span>
<span class="n">cre_seq</span> <span class="o">=</span> <span class="n">input_seq</span><span class="p">[</span><span class="n">cre_pos</span><span class="p">:</span> <span class="n">cre_pos</span><span class="o">+</span><span class="mi">5000</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="sufficiency-test">
<h3>2.3 Sufficiency test<a class="headerlink" href="#sufficiency-test" title="Link to this heading">#</a></h3>
<p>The sufficiency test (also from CREME) consists of the following steps:</p>
<ol class="arabic simple">
<li><p>Predict TSS activity for the wild type sequence (WT).</p></li>
<li><p>Dinucleotide shuffle the sequence.</p></li>
<li><p>Add the TSS 5 kb tile and predict TSS activity (CONTROL).</p></li>
<li><p>Add the CRE and the TSS tiles to the sequence and predict TSS activity (MUTANT).</p></li>
<li><p>in case of enhancing context sequences, we compute the normalized CRE effect as (MUTANT - CONTROL) / WT.</p></li>
<li><p>Repeat each shuffle 10 times and average the normalized CRE effect per sequence.</p></li>
</ol>
<p>We insert the CRE into our 10 shuffled sequences containing the TSS.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">shuffled_seqs_w_tss_and_cre</span> <span class="o">=</span> <span class="p">[</span><span class="n">mutate</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">cre_seq</span><span class="p">,</span> <span class="n">pos</span><span class="o">=</span><span class="n">cre_pos</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">shuf_seqs_w_tss</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<p>Note that what was previously called MUTANT (shuffled sequence with the TSS tile) is now called CONTROL, so we rename it:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">CONTROL</span> <span class="o">=</span> <span class="n">MUTANT</span>
<span class="k">del</span> <span class="n">MUTANT</span>
</pre></div>
</div>
</div>
</div>
<p>And then make predictions:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ds</span> <span class="o">=</span> <span class="n">SeqDataset</span><span class="p">(</span><span class="n">shuffled_seqs_w_tss_and_cre</span><span class="p">)</span>
<span class="n">MUTANT</span> <span class="o">=</span> <span class="n">enformer</span><span class="o">.</span><span class="n">predict_on_dataset</span><span class="p">(</span><span class="n">ds</span><span class="p">,</span> <span class="n">devices</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">batch_size</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span><span class="o">.</span><span class="n">squeeze</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Predicting DataLoader 0: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00&lt;00:00,  2.59it/s]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">MUTANT</span> <span class="o">=</span> <span class="n">tss_activity_transform</span><span class="o">.</span><span class="n">compute</span><span class="p">(</span><span class="n">MUTANT</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">results</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">((</span><span class="n">MUTANT</span> <span class="o">-</span> <span class="n">CONTROL</span><span class="p">)</span> <span class="o">/</span> <span class="n">WT</span><span class="p">)</span>
<span class="n">results</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>0.02895034
</pre></div>
</div>
</div>
</div>
<p>This suggests that although shuffling the CRE has a large effect on TSS activity, it is not sufficient to increase TSS activity much on its own.</p>
</section>
<section id="tss-cre-distance-test">
<h3>2.4 TSS-CRE Distance Test<a class="headerlink" href="#tss-cre-distance-test" title="Link to this heading">#</a></h3>
<p>CREME defined a ‘TSS-CRE Distance test’ in which the TSS is inserted into a dinucleotide-shuffled sequence, the putative CRE is inserted at different distances from the TSS, and its effect is measured at each distance. We can do a similar experiment using the previously seen <code class="docutils literal notranslate"><span class="pre">marginalize_pattern_spacing</span></code> function.</p>
<p>Note that CREME normalizes effects using the maximum effect of the CRE at any difference, which we do not do here, but can be done manually.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">preds</span><span class="p">,</span> <span class="n">distances</span> <span class="o">=</span> <span class="n">grelu</span><span class="o">.</span><span class="n">interpret</span><span class="o">.</span><span class="n">simulate</span><span class="o">.</span><span class="n">marginalize_pattern_spacing</span><span class="p">(</span>
    <span class="n">model</span> <span class="o">=</span> <span class="n">enformer</span><span class="p">,</span>
    <span class="n">seqs</span> <span class="o">=</span> <span class="n">input_seq</span><span class="p">,</span>
    <span class="n">fixed_pattern</span> <span class="o">=</span> <span class="n">tss_seq</span><span class="p">,</span> <span class="c1"># The TSS will be fixed in the center of the sequence</span>
    <span class="n">moving_pattern</span> <span class="o">=</span> <span class="n">cre_seq</span><span class="p">,</span> <span class="c1"># The CRE will be inserted at different distances</span>
    <span class="n">stride</span><span class="o">=</span><span class="mi">5000</span><span class="p">,</span> <span class="c1"># The CRE insertion position will shift by 5000 bp</span>
    <span class="n">n_shuffles</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span> <span class="c1"># The input sequence will be shuffled 10 times and results averaged</span>
    <span class="n">seed</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
    <span class="n">devices</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
    <span class="n">num_workers</span><span class="o">=</span><span class="mi">16</span><span class="p">,</span>
    <span class="n">batch_size</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span>
    <span class="n">prediction_transform</span><span class="o">=</span><span class="n">tss_activity_transform</span><span class="p">,</span>
    <span class="n">compare_func</span><span class="o">=</span><span class="s1">&#39;subtract&#39;</span><span class="p">,</span> <span class="c1"># Compute the change in prediction after inserting the CRE. </span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Predicting DataLoader 0: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 48/48 [00:26&lt;00:00,  1.80it/s]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">preds</span> <span class="o">=</span> <span class="n">preds</span><span class="o">.</span><span class="n">squeeze</span><span class="p">()</span>
<span class="n">preds</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(10, 37)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">len</span><span class="p">(</span><span class="n">distances</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>37
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">grelu</span><span class="o">.</span><span class="n">visualize</span><span class="o">.</span><span class="n">plot_position_effect</span><span class="p">(</span>
    <span class="n">preds</span><span class="p">,</span> <span class="n">distances</span><span class="p">,</span>
    <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Change in predicted TSS activity</span><span class="se">\n</span><span class="s2">after inserting enhancer&quot;</span><span class="p">,</span>
    <span class="n">xlab</span><span class="o">=</span><span class="s2">&quot;Start position of tile&quot;</span><span class="p">,</span>
    <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span><span class="mi">2</span><span class="p">)</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="../_images/a70b77b5f6c99d4b537a2feb17a13f9705afc5901d68f79a03e365d5e2438a9e.png"><img alt="../_images/a70b77b5f6c99d4b537a2feb17a13f9705afc5901d68f79a03e365d5e2438a9e.png" src="../_images/a70b77b5f6c99d4b537a2feb17a13f9705afc5901d68f79a03e365d5e2438a9e.png" style="width: 686px; height: 288px;" /></a>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;distance&#39;</span><span class="p">:</span><span class="n">distances</span><span class="p">,</span> <span class="s1">&#39;effect&#39;</span><span class="p">:</span> <span class="n">preds</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="mi">0</span><span class="p">)})</span>
<span class="n">df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s1">&#39;effect&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>distance</th>
      <th>effect</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>15</th>
      <td>-20804</td>
      <td>0.591190</td>
    </tr>
    <tr>
      <th>16</th>
      <td>-15804</td>
      <td>0.558111</td>
    </tr>
    <tr>
      <th>21</th>
      <td>19196</td>
      <td>0.536405</td>
    </tr>
    <tr>
      <th>20</th>
      <td>14196</td>
      <td>0.507782</td>
    </tr>
    <tr>
      <th>17</th>
      <td>-10804</td>
      <td>0.491934</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
</section>
</section>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#part-1-catlas-atac-seq-binary-classification-model">Part 1: CATlas ATAC-seq binary classification model</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#motif-marginalization">1.1 Motif marginalization</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#motif-spacing">1.2 Motif spacing</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#part-2-enformer">Part 2: Enformer</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#context-dependence-test">2.1 Context dependence test</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#enhancer-discovery-by-in-silico-tiling">2.2 Enhancer Discovery by in silico tiling</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sufficiency-test">2.3 Sufficiency test</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#tss-cre-distance-test">2.4 TSS-CRE Distance Test</a></li>
</ul>
</li>
</ul>
  </nav></div>

  <div class="sidebar-secondary-item">

  
  <div class="tocsection editthispage">
    <a href="https://github.com/Genentech/gReLU/edit/main/docs/tutorials/7_simulations.ipynb">
      <i class="fa-solid fa-pencil"></i>
      
      
        
          Edit on GitHub
        
      
    </a>
  </div>
</div>

  <div class="sidebar-secondary-item">
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/tutorials/7_simulations.ipynb"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2023, Regulatory Genomics, BRAID, Genentech.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.2.0.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>